pallets/messages/src/types.rs

```diff
@@ -40,3 +40,43 @@ where
 		}
 	}
 }
+
+/// Newtype for CIDv2 addresses. We can change the inner type as we understand
+/// more about how to structure IPFS addressable content.
+#[derive(Clone, Encode, Decode, PartialEq, Debug, TypeInfo, Eq)]
+pub struct CIDv2(Vec<u8>);
+
+impl CIDv2 {
+	/// Constructor
+	pub fn new(vec: Vec<u8>) -> CIDv2 {
+		CIDv2(vec)
+	}
+
+	/// Exposes the inner vector
+	pub fn get(&self) -> &Vec<u8> {
+		&self.0
+	}
+}
+
+/// Offchain payloads are defined by two members, a CID address and a payload length
+pub struct OffchainPayload {
+	/// IPFS Content address
+	pub cid: CIDv2,
+	/// Size of the IPFS payload
+	pub payload_length: u32,
+}
+
+impl OffchainPayload {
+	/// Constructor
+	pub fn new(cid: CIDv2, payload_length: u32) -> OffchainPayload {
+		OffchainPayload { cid, payload_length }
+	}
+}
+
+/// Sum type for payloads
+pub enum Payload {
```

Comment:
I think that _might_ end up making things confusing. The `PayloadLocation` is used as a an attribute of the schema, whereas the payload is the data described by the schema.

In theory we could parametrize the `PayloadLocation` enum, but we would still need to store the `CID` and payload length in another struct, since they could differ within a single schema definition.